release/train_tests/benchmark/recsys/torchrec_runner.py

```diff
@@ -0,0 +1,138 @@
+import logging
+import gc
+import os
+
+import torch
+import torch.nn
+
+from torchrec.distributed.train_pipeline import StagedTrainPipeline, SparseDataDistUtil
+from torchrec.distributed.train_pipeline.utils import PipelineStage
+from torchrec.optim.keyed import CombinedOptimizer, KeyedOptimizerWrapper
+from torchrec.optim.optimizers import in_backward_optimizer_filter
+
+import ray.train
+import ray.train.torch
+
+from runner import TrainLoopRunner
+
+
+logger = logging.getLogger(__name__)
+
+
+class TorchRecRunner(TrainLoopRunner):
+    def _setup(self):
+        if self.factory.benchmark_config.mock_gpu:
+            raise ValueError("Mock GPU is not supported for running TorchRec.")
+
+        self.model = self.factory.get_model()
+
+        # TODO: This code depends on the model having a fused_optimizer,
+        # which is hidden in the `get_model` method of the factory.
+        dense_optimizer = KeyedOptimizerWrapper(
+            dict(in_backward_optimizer_filter(self.model.named_parameters())),
+            lambda params: torch.optim.Adagrad(params, lr=15.0, eps=1e-8),
+        )
+        self.optimizer = CombinedOptimizer(
+            [self.model.fused_optimizer, dense_optimizer]
+        )
+
+        self._data_dist_stream = torch.cuda.Stream()
+        self._h2d_stream = torch.cuda.Stream()
+
+    def _wrap_dataloader(self, dataloader, train: bool = True):
+        dataloader_iter = iter(dataloader)
+
+        device = ray.train.torch.get_device()
+        sdd = SparseDataDistUtil(
+            model=self.model,
+            data_dist_stream=self._data_dist_stream,
+            # prefetch_stream=torch.cuda.Stream(),
+        )
+        pipeline = [
+            PipelineStage(
+                name="data_copy",
+                runnable=lambda batch: batch.to(device, non_blocking=True),
+                stream=self._h2d_stream,
+            ),
+            PipelineStage(
+                name="start_sparse_data_dist",
+                runnable=sdd.start_sparse_data_dist,
+                stream=sdd.data_dist_stream,
+                fill_callback=sdd.wait_sparse_data_dist,
+            ),
+            # PipelineStage(
+            #     name="prefetch",
+            #     runnable=sdd.prefetch,
+            #     stream=sdd.prefetch_stream,
+            #     fill_callback=sdd.load_prefetch,
+            # ),
+        ]
+
+        pipeline = StagedTrainPipeline(pipeline_stages=pipeline)
+
+        def dataloader_with_torchrec_pipeline():
+            while batch := pipeline.progress(dataloader_iter):
+                yield batch
+
+            pipeline.flush_end()
```

Comment:
This is the torchrec utility to pipeline the device transfer, data shuffling, and model forward/backward steps.